### PR TITLE
Digifinex: swap support

### PR DESCRIPTION
**Pull request by Dan-krm**
Added the remaining swap support to Digifinex:
### fetchBalance:
```
node examples/js/cli digifinex fetchBalance '{"type":"swap"}'

{
  info: [
    {
      equity: '0',
      currency: 'BTC',
      margin: '0',
      cross_margin: '0',
      isolate_margin: '0',
      isolate_balance: '0',
      frozen_margin: '0',
      frozen_money: '0',
      margin_ratio: '0',
      realized_pnl: '0',
      avail_balance: '0',
      unrealized_pnl: '0',
      isolate_unrealized_pnl: '',
      time_stamp: '1661487402396'
    },
...
  ],
  BTC: { free: 0, used: 0, total: 0 },
  DOT: { free: 0, used: 0, total: 0 },
  EOS: { free: 0, used: 0, total: 0 },
...
  free: { BTC: 0, DOT: 0, EOS: 0, LINK: 0, TRX: 0, USDT: 21.4512422 },
  used: { BTC: 0, DOT: 0, EOS: 0, LINK: 0, TRX: 0, USDT: 0 },
  total: { BTC: 0, DOT: 0, EOS: 0, LINK: 0, TRX: 0, USDT: 21.4512422 }
}
```

### fetchOrderBook:
```
node examples/js/cli digifinex fetchOrderBook BTC/USDT:USDT

{
  symbol: 'BTC/USDT:USDT',
  bids: [
    [ 16122.8, 3027 ], [ 16122.3, 3193 ],
    [ 16121.8, 3093 ], [ 16121.7, 3115 ],
    [ 16121.6, 3391 ], [ 16121.5, 3068 ],
...
  ],
  asks: [
    [ 16122.9, 3488 ], [ 16123.5, 2826 ],
    [ 16123.8, 3379 ], [ 16123.9, 3228 ],
    [ 16124, 3310 ],   [ 16124.1, 3032 ],
...
  ],
  timestamp: 1669155919547,
  datetime: '2022-11-22T22:25:19.547Z',
  nonce: undefined
}
```

### fetchTrades:
```
node examples/js/cli digifinex fetchTrades BTC/USDT:USDT

                 id |     timestamp |                 datetime |        symbol | type | order |        side |   price | amount |       cost | takerOrMaker | fee | fees
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
1595192330640429057 | 1669158463522 | 2022-11-22T23:07:43.522Z | BTC/USDT:USDT |      |       |  close long | 16197.2 |     30 |    485.916 |              |     |   []
1595192332464951298 | 1669158463957 | 2022-11-22T23:07:43.957Z | BTC/USDT:USDT |      |       | close short | 16197.3 |    411 |  6657.0903 |              |     |   []
1595192334075564033 | 1669158464341 | 2022-11-22T23:07:44.341Z | BTC/USDT:USDT |      |       | close short | 16198.3 |    177 |  2867.0991 |              |     |   []
...
20 objects
```

### fetchOHLCV:
```
node examples/js/cli digifinex fetchOHLCV BTC/USDT:USDT

1669180620000 | 16495.9 | 16499.8 | 16494.3 | 16494.4 |  9405
1669180680000 | 16494.6 |   16503 | 16494.6 | 16495.1 | 13784
1669180740000 | 16495.1 | 16496.7 | 16484.1 | 16484.7 | 17481
...
20 objects
```

### fetchMyTrades:
```
node examples/js/cli digifinex fetchMyTrades BTC/USDT:USDT

                 id |     timestamp |                 datetime |        symbol | type |               order |       side |   price | amount |     cost | takerOrMaker |                 fee |                  fees
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
1590136768424841218 | 1667953123562 | 2022-11-09T00:18:43.562Z | BTC/USDT:USDT |      | 1590136768156405760 |  open long | 18514.5 |      1 |  18.5145 |        taker | {"cost":0.00925725} | [{"cost":0.00925725}]
1590136799731126273 | 1667953131026 | 2022-11-09T00:18:51.026Z | BTC/USDT:USDT |      | 1590136799198449664 | close long | 18515.5 |      1 |  18.5155 |        taker | {"cost":0.00925775} | [{"cost":0.00925775}]
1590165683381997569 | 1667960017426 | 2022-11-09T02:13:37.426Z | BTC/USDT:USDT |      | 1590165683096784896 |  open long | 18369.3 |      1 |  18.3693 |        taker | {"cost":0.00918465} | [{"cost":0.00918465}]
...
20 objects
```

### fetchLedger:
```
node examples/js/cli digifinex fetchLedger undefined undefined undefined '{"type":"swap"}'

id | direction | account | referenceId | referenceAccount | type | currency |      amount | before | after | status |     timestamp |                 datetime | fee
--------------------------------------------------------------------------------------------------------------------------------------------------------------------
   |           |         |             |                  |   17 |     USDT |        3.01 |        |       |        | 1650809432000 | 2022-04-24T14:10:32.000Z |

   |           |         |             |                  |   20 |     USDT |       -3.01 |        |       |        | 1652022010000 | 2022-05-08T15:00:10.000Z |

   |           |         |             |                  |   17 |     USDT |        2.01 |        |       |        | 1652071979000 | 2022-05-09T04:52:59.000Z |

...
14 objects
```

### fetchLeverageTiers:
```
node examples/js/cli digifinex fetchLeverageTiers '["EGLD/USDT:USDT","UNI/USD:UNI"]'

{
  'EGLD/USDT:USDT': [
    {
      tier: 1,
      currency: 'USDT',
      minNotional: undefined,
      maxNotional: 200000,
      maintenanceMarginRate: undefined,
      maxLeverage: 3,
      info: { leverage: '3', max_limit: '200000' }
    },
    {
      tier: 2,
      currency: 'USDT',
      minNotional: undefined,
      maxNotional: 100000,
      maintenanceMarginRate: undefined,
      maxLeverage: 10,
      info: { leverage: '10', max_limit: '100000' }
    },
    {
      tier: 3,
      currency: 'USDT',
      minNotional: undefined,
      maxNotional: 30000,
      maintenanceMarginRate: undefined,
      maxLeverage: 50,
      info: { leverage: '50', max_limit: '30000' }
    }
  ],
  'UNI/USD:UNI': [
    {
      tier: 1,
      currency: 'UNI',
      minNotional: undefined,
      maxNotional: 20000,
      maintenanceMarginRate: undefined,
      maxLeverage: 3,
      info: { leverage: '3', max_limit: '20000' }
    },
    {
      tier: 2,
      currency: 'UNI',
      minNotional: undefined,
      maxNotional: 10000,
      maintenanceMarginRate: undefined,
      maxLeverage: 10,
      info: { leverage: '10', max_limit: '10000' }
    },
    {
      tier: 3,
      currency: 'UNI',
      minNotional: undefined,
      maxNotional: 5000,
      maintenanceMarginRate: undefined,
      maxLeverage: 50,
      info: { leverage: '50', max_limit: '5000' }
    }
  ]
}
```

### fetchMarketLeverageTiers:
```
node examples/js/cli digifinex fetchMarketLeverageTiers EGLD/USDT:USDT

tier | currency | minNotional | maxNotional | maintenanceMarginRate | maxLeverage
---------------------------------------------------------------------------------
   1 |     USDT |             |      200000 |                       |           3
   2 |     USDT |             |      100000 |                       |          10
   3 |     USDT |             |       30000 |                       |          50
3 objects
```

### CODE REVIEW

**Comment by samgermain** [comment]
splitting this into multiple PRs will make it easier to merge

File: js/digifinex.js [hunk 1]
```diff
@@ -41,11 +42,18 @@ module.exports = class digifinex extends Exchange {
                 'fetchCurrencies': true,
                 'fetchDepositAddress': true,
                 'fetchDeposits': true,
+                'fetchFundingHistory': false,
```
Comment (samgermain) [inline js/digifinex.js:45]: I think this endpoint might provide it https://docs.bitfinex.com/reference/ws-auth-funding-trades

Comment (Dan-krm) [reply js/digifinex.js:45]: That's the wrong exchange

File: js/digifinex.js [hunk 2]
```diff
@@ -1031,23 +1137,79 @@ module.exports = class digifinex extends Exchange {
...
+            }
+            if (tradeRole === '1') {
+                takerOrMaker = 'taker';
+            } else if (tradeRole === '2') {
+                takerOrMaker = 'maker';
+            } else {
+                takerOrMaker = undefined;
+            }
+            if ((side === '1') || (direction === '1')) {
+                side = 'open long';
```
Comment (carlosmiei) [inline js/digifinex.js:1187]: @Dan-krm, these are more informative for sure, but shouldn't we be using the unified side here? (buy/sell)

File: js/digifinex.js [hunk 3]
```diff
@@ -1031,23 +1137,79 @@ module.exports = class digifinex extends Exchange {
...
+            }
+        } else {
+            const parts = side.split ('_');
+            side = this.safeString (parts, 0);
+            type = this.safeString (parts, 1);
+            if (type === undefined) {
+                type = 'limit';
+            }
+            const isMaker = this.safeValue (trade, 'is_maker');
+            takerOrMaker = (isMaker) ? 'maker' : 'taker';
```
Comment (carlosmiei) [inline js/digifinex.js:1203]: remove these extra () pls

**Comment by carlosmiei** [comment]
LGTM ✅ 

```
p digifinex fetchOrderBook "BTC/USDT" 1               
Python v3.10.8
CCXT v2.2.29
digifinex.fetchOrderBook(BTC/USDT,1)
{'asks': [[16359.79, 0.03680237]],
 'bids': [[16359.7, 0.084185]],
 'datetime': '2022-11-29T15:14:14.000Z',
 'nonce': None,
 'symbol': 'BTC/USDT',
 'timestamp': 1669734854000}
 ```

```
 digifinex fetchOrderBook "BTC/USDT:USDT" 1
Python v3.10.8
CCXT v2.2.29
digifinex.fetchOrderBook(BTC/USDT:USDT,1)
{'asks': [[16366.3, 2098.0]],
 'bids': [[16366.2, 10856.0]],
 'datetime': '2022-11-29T15:14:29.283Z',
 'nonce': None,
 'symbol': 'BTC/USDT:USDT',
 'timestamp': 1669734869283}
 ```

```
p digifinex fetchTrades "BTC/USDT" None 1   
Python v3.10.8
CCXT v2.2.29
digifinex.fetchTrades(BTC/USDT,None,1)
[{'amount': 0.05697113,
  'cost': 932.6749389413,
  'datetime': '2022-11-29T15:14:52.000Z',
  'fee': None,
  'fees': [],
  'id': '12898869602',
  'info': {'amount': '0.05697113',
           'date': '1669734892',
           'id': '12898869602',
           'price': '16371.01',
           'type': 'sell'},
  'order': None,
  'price': 16371.01,
  'side': 'sell',
  'symbol': 'BTC/USDT',
  'takerOrMaker': 'taker',
  'timestamp': 1669734892000,
  'type': 'limit'}]
  ```

```
p digifinex fetchTrades "BTC/USDT:USDT" None 1
Python v3.10.8
CCXT v2.2.29
digifinex.fetchTrades(BTC/USDT:USDT,None,1)
[{'amount': 82.0,
  'cost': 1342.5204,
  'datetime': '2022-11-29T15:15:16.846Z',
  'fee': None,
  'fees': [],
  'id': '1597610151123095553',
  'info': {'direction': '4',
           'instrument_id': 'BTCUSDTPERP',
           'price': '16372.2',
           'trade_id': '1597610151123095553',
           'trade_time': '1669734916846',
           'volume': '82'},
  'order': None,
  'price': 16372.2,
  'side': 'close short',
  'symbol': 'BTC/USDT:USDT',
  'takerOrMaker': None,
  'timestamp': 1669734916846,
  'type': None}]
  ```

```
p digifinex fetchOHLCV "BTC/USDT" 1m None 1
Python v3.10.8
CCXT v2.2.29
digifinex.fetchOHLCV(BTC/USDT,1m,None,1)
[[1669735500000, 16395.2, 16396.69, 16389.21, 16389.6, 3.5785374]]
```

```
 p digifinex fetchOHLCV "BTC/USDT:USDT" 1m None 1
Python v3.10.8
CCXT v2.2.29
digifinex.fetchOHLCV(BTC/USDT:USDT,1m,None,1)
[[1669735560000, 16384.2, 16384.2, 16384.1, 16384.1, 957.0]]
```

```
p digifinex fetchMyTrades "LTC/USDT:USDT"
Python v3.10.8
CCXT v2.2.29
digifinex.fetchMyTrades(LTC/USDT:USDT)
[{'amount': 3.0,
  'cost': 17.6934,
  'datetime': '2022-11-17T10:29:27.322Z',
  'fee': {'cost': 0.0088467, 'currency': None},
  'fees': [{'cost': 0.0088467, 'currency': None}],
  'id': '1593189566418784257',
  'info': {'close_profit': '0',
           'fee': '0.0088467',
           'instrument_id': 'LTCUSDTPERP',
           'leverage': '5',
           'margin_mode': 'crossed',
           'match_role': '1',
           'order_id': '1593189566343286784',
           'order_type': '8',
           'price': '58.978',
           'size': '3',
           'trade_id': '1593189566418784257',
           'trade_time': '1668680967322',
           'trade_type': '0',
           'type': '1'},
  'order': '1593189566343286784',
  'price': 58.978,
  'side': 'open long',
  'symbol': 'LTC/USDT:USDT',
  'takerOrMaker': 'taker',
  'timestamp': 1668680967322,
  'type': None}]
  ```

```
 p digifinex fetchMyTrades "LTC/USDT:USDT"
Python v3.10.8
CCXT v2.2.29
digifinex.fetchMyTrades(LTC/USDT:USDT)
[{'amount': 3.0,
  'cost': 17.6934,
  'datetime': '2022-11-17T10:29:27.322Z',
  'fee': {'cost': 0.0088467, 'currency': None},
  'fees': [{'cost': 0.0088467, 'currency': None}],
  'id': '1593189566418784257',
  'info': {'close_profit': '0',
           'fee': '0.0088467',
           'instrument_id': 'LTCUSDTPERP',
           'leverage': '5',
           'margin_mode': 'crossed',
           'match_role': '1',
           'order_id': '1593189566343286784',
           'order_type': '8',
           'price': '58.978',
           'size': '3',
           'trade_id': '1593189566418784257',
           'trade_time': '1668680967322',
           'trade_type': '0',
           'type': '1'},
  'order': '1593189566343286784',
  'price': 58.978,
  'side': 'buy',
  'symbol': 'LTC/USDT:USDT',
  'takerOrMaker': 'taker',
  'timestamp': 1668680967322,
  'type': None}]
  ```

**Review by carlosmiei — APPROVED** [review]

**Pull request merged**